pkg/vkMachinery/csr/generation_test.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2019-2021 The Liqo Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package csr
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"net"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Assert the correctness of the virtual-node CSR", func() {
+	var (
+		podNameT string
+		podIP    net.IP
+		csrPEM   []byte
+		keyPEM   []byte
+		csr      *x509.CertificateRequest
+	)
+	When("Requesting a certificate for a virtual node", func() {
+		BeforeEach(func() {
+			podNameT = "podName"
+			podIP = net.ParseIP("10.0.0.1")
+		})
+
+		JustBeforeEach(func() {
+			csrPEM, keyPEM, err = generateVKCertificateBundle(podNameT, podIP)
+		})
+
+		It("Should not trigger any error", func() {
+			Expect(err).NotTo(HaveOccurred())
+			Expect(csrPEM).NotTo(BeNil())
+			Expect(keyPEM).NotTo(BeNil())
+		})
+
+		It("Should be a valid x509 certificate", func() {
+			b, _ := pem.Decode(csrPEM)
+			Expect(b).NotTo(BeNil())
+			csr, err = x509.ParseCertificateRequest(b.Bytes)
+			Expect(err).NotTo(HaveOccurred())
```

Comment:
Side-effects should be performed in BeforeEach blocks, since you should not rely on It block ordering. Here, you could create a nested Context block, in the BeforeEach you decode the CSR, and then the It blocks check the correctness.